backends/s3/s3.go

```diff
@@ -189,9 +182,38 @@ func (b *Backend) Store(ctx context.Context, name string, data []byte) error {
 		return err
 	}
 	if b.opt.UseUpdateMarker {
+		var wg sync.WaitGroup
+		wg.Add(1)
+
+		// Update cache
+		go func() {
+			defer wg.Done()
+			// Update size or add to local list if not present
+			l := b.lastList.Len()
+			idx := sort.Search(l, func(i int) bool { return b.lastList[i].Name >= name })
+			blob := simpleblob.Blob{Name: name, Size: int64(len(data))}
+			b.mu.Lock()
+			if idx < l {
+				if b.lastList[idx].Name == name {
+					b.lastList[idx].Size = int64(len(data))
+				} else {
+					b.lastList = append(b.lastList, simpleblob.Blob{})
+					copy(b.lastList[idx+1:], b.lastList[idx:])
```

Comment:
The whole clever insert causes race conditions with locking.

If you always set the attributes to a completely new value, it is safe to return that value and use access it when not locked, which is what we do.

With the insertion trick, the values are no longer immutable and can no longer be accessed without holding a lock.

In the old version `b.lastList` could safely be returned to the caller and the caller could access the items without holding a lock.

In the new version accessing the items in the slice without holding the lock (an implementation detail) can/will cause race conditions when the list is updated at the same time.
 
Instead of the insert, create a new slice and then assign the new slide to `b.lastList` in a tiny critical section. Something like:

```go
b.mu.Lock()
oldList := b.lastList 
b.mu.Unlock()

var newList []simpleblob.Blob
// fill newList here

b.mu.Lock()
b.lastList = newList
b.mu.Unlock()
```